ethstats/ethstats.go

```diff
@@ -211,93 +253,215 @@ func (s *Service) loop() {
+	Info    nodeInfo       `json:"info"`
+}
 
-			case <-fullReport.C:
-				if err = s.report(conn); err != nil {
-					log.Warn("Full stats report failed", "err", err)
-				}
-			case list := <-s.histCh:
-				if err = s.reportHistory(conn, list); err != nil {
-					log.Warn("Requested history report failed", "err", err)
-				}
-			case head := <-headCh:
-				if err = s.reportBlock(conn, head); err != nil {
-					log.Warn("Block stats report failed", "err", err)
-				}
-				if err = s.reportPending(conn); err != nil {
-					log.Warn("Post-block transaction stats report failed", "err", err)
-				}
-			case <-txCh:
-				if err = s.reportPending(conn); err != nil {
-					log.Warn("Transaction stats report failed", "err", err)
-				}
+// login tries to authorize the client at the remote server.
+func (s *Service) login(conn *websocket.Conn, sendCh chan StatsPayload) error {
+	// Construct and send the login authentication
+	infos := s.server.NodeInfo()
+
+	var (
+		etherBase common.Address
+		network   string
+		protocol  string
+	)
+	p := s.engine.Protocol()
+	if info := infos.Protocols[p.Name]; info != nil {
+		network = fmt.Sprintf("%d", info.(*eth.NodeInfo).Network)
+		protocol = fmt.Sprintf("%s/%d", p.Name, p.Versions[0])
+	} else {
+		network = fmt.Sprintf("%d", infos.Protocols["les"].(*les.NodeInfo).Network)
+		protocol = fmt.Sprintf("les/%d", les.ClientProtocolVersions[0])
+	}
+	if s.eth != nil {
+		etherBase, _ = s.eth.Etherbase()
+	}
+	auth := &authMsg{
+		ID:      s.node,
+		Address: etherBase,
+		Info: nodeInfo{
+			Name:     s.node,
+			Node:     infos.Name,
+			Port:     infos.Ports.Listener,
+			Network:  network,
+			Protocol: protocol,
+			API:      "No",
+			Os:       runtime.GOOS,
+			OsVer:    runtime.GOARCH,
+			Client:   "0.1.1",
+			History:  true,
+		},
+	}
+	if err := s.sendStats(conn, actionHello, auth); err != nil {
+		return err
+	}
+
+	// Proxy needs a delegate send here to get ACK
+	if s.backend.ProxiedPeer() != nil {
+		select {
+		case signedMessage := <-sendCh:
+			err := s.handleDelegateSend(conn, signedMessage)
+			if err != nil {
+				return err
 			}
+		case <-time.After(5 * time.Second):
+			// Login timeout, abort
+			return errors.New("login timed out")
+		}
+	}
+
+	// Retrieve the remote ack or connection termination
+	var ack map[string][]string
+	if err := websocket.JSON.Receive(conn, &ack); err != nil || len(ack["emit"]) != 1 || ack["emit"][0] != "ready" {
+		return errors.New("unauthorized")
+	}
+	return nil
+}
+
+func (s *Service) handleDelegateSign(messageToSign StatsPayload) error {
+
+	if s.backend.ProxyNode() == nil {
+		return errors.New("No Proxy found")
+	}
+
+	signedStats, err := s.signStats(messageToSign.Stats)
+
+	if err == nil {
+		signedMessage := &StatsPayload{
+			Action: messageToSign.Action,
+			Stats:  signedStats,
 		}
-		// Make sure the connection is closed
-		conn.Close()
+		msg, _ := json.Marshal(signedMessage)
+		return s.backend.ProxyNode().Send(0x15, msg)
```

Comment:
If we do sending in here, should we do it as a goroutine? We also shouldn't hardcode `0x15`